pkg/koordlet/runtimehooks/nri/server.go

```diff
@@ -0,0 +1,226 @@
+			} else {
+				klog.Info("nri server start")
+			}
+		} else {
+			klog.Error("nri stub is nil")
+		}
+	}()
+	return nil
+}
+
+func (p *NriServer) Configure(config, runtime, version string) (stub.EventMask, error) {
+	klog.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	err := yaml.Unmarshal([]byte(config), &cfg)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse provided configuration: %w", err)
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	return p.mask, nil
+}
+
+func (p *NriServer) Synchronize(pods []*api.PodSandbox, containers []*api.Container) ([]*api.ContainerUpdate, error) {
+	// todo: update existed containers configure
+	return nil, nil
+}
+
+func (p *NriServer) RunPodSandbox(pod *api.PodSandbox) error {
+	podCtx := &protocol.PodContext{}
+	podCtx.FromNri(pod)
+	// todo: return error or bypass error based on PluginFailurePolicy
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreRunPodSandbox, podCtx)
+	if err != nil {
+		klog.Errorf("nri hooks run error: %v", err)
+		if p.options.PluginFailurePolicy == rmconfig.PolicyFail {
+			return err
+		}
+	}
+	podCtx.NriDone(p.options.Executor)
+	return nil
+}
+
+func (p *NriServer) CreateContainer(pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	containerCtx := &protocol.ContainerContext{}
+	containerCtx.FromNri(pod, container)
+	// todo: return error or bypass error based on PluginFailurePolicy
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreCreateContainer, containerCtx)
+	if err != nil {
+		klog.Errorf("nri run hooks error: %v", err)
+		if p.options.PluginFailurePolicy == rmconfig.PolicyFail {
+			return nil, nil, err
+		}
+	}
+
+	adjust, _, err := Protocol2NRI(containerCtx)
+	if err != nil {
+		klog.Errorf("nri convert protocol to : %v", err)
+		return nil, nil, nil
+	}
+	return adjust, nil, nil
+}
+
+func (p *NriServer) UpdateContainer(pod *api.PodSandbox, container *api.Container) ([]*api.ContainerUpdate, error) {
+	containerCtx := &protocol.ContainerContext{}
+	containerCtx.FromNri(pod, container)
+	// todo: return error or bypass error based on PluginFailurePolicy
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreCreateContainer, containerCtx)
+	if err != nil {
+		klog.Errorf("nri run hooks error: %v", err)
+		if p.options.PluginFailurePolicy == rmconfig.PolicyFail {
+			return nil, err
+		}
+	}
+
+	_, update, err := Protocol2NRI(containerCtx)
+	if err != nil {
+		klog.Errorf("nri convert protocol to : %v", err)
+		return nil, nil
+	}
+
+	return []*api.ContainerUpdate{update}, nil
+}
+
+func (p *NriServer) onClose() {
+	p.stub.Stop()
+}
+
+func Protocol2NRI(proto protocol.HooksProtocol) (*api.ContainerAdjustment, *api.ContainerUpdate, error) {
```

Comment:
We may need a TODO here to add more field conversions (e.g. `memory.min` in cgroups-v2).